Port MASTG-TEST-0088 (by @appknox) (#3073)
OWASP/owasp-mastg: tests-beta/ios/MASVS-RESILIENCE/MASTG-TEST-0x89.md

## Steps

1. Run a dynamic analysis tool such as @MASTG-TOOL-0038 on the app and execute its [automated jailbreak detection bypass commands](../../../Document/0x06j-Testing-Resiliency-Against-Reverse-Engineering.md#automated-jailbreak-detection-bypass)

## Observation

The output should include any instances of common jailbreak detection checks in the app binary and the results of the automated jailbreak detection bypass commands.

## Evaluation

The test passes if jailbreak detection is implemented and the automated jailbreak detection bypass commands are successful or at least provide useful information to help you narrow down the jailbreak detection checks implemented in the app.

The test fails if jailbreak detection is not implemented. However, note that this test is not exhaustive and may not detect all jailbreak detection checks as it relies on predefined bypass code that may not cover all possible jailbreak detection checks or may not be up-to-date. The checks may also be more sophisticated than what the tool can detect so manual reverse engineering and deobfuscation may be required to identify them.

> 

Reviewer: ```suggestion
```

---
platform: ios
title: Runtime Use of Jailbreak Detection Techniques
id: MASTG-TEST-0x89
type: [dynamic]
weakness: MASWE-0097
false_negative_prone: true
---

## Overview

> The test verifies that a mobile application can identify if the iOS device it is running on is jailbroken. It does so by dynamically analyzing the app binary for [common jailbreak detection checks](../../../Document/0x06j-Testing-Resiliency-Against-Reverse-Engineering.md#common-jailbreak-detection-checks) and trying to bypass them. For example, it may detect a check for the presence of certain files or directories that are indicative of a jailbroken device and use dynamic instrumentation to bypass them by modifying the list of files or directories that are checked.

Reviewer 2: ```suggestion
The test verifies that a mobile application can identify if the iOS device it is running on is jailbroken. It does so by dynamically analyzing the app binary for [common jailbreak detection checks](../../../Document/0x06j-Testing-Resiliency-Against-Reverse-Engineering.md#common-jailbreak-detection-checks) and trying to bypass them. For example, it may detect a check for the presence of certain files or directories that are indicative of a jailbroken device.
```
Removed the last sentence as the bypass angle is irrelevant. We're using the tool to collect artefacts, not to bypass a JB check

The limitations of using jailbreak detection bypass tools should be considered. It is possible that the app uses more sophisticated jailbreak detection techniques that are not detected by the tool used. In such cases, careful manual reverse engineering and deobfuscation are required to identify the jailbreak detection checks. Also, additional dynamic analysis work may be required to bypass the jailbreak detection checks.

## Steps

1. Run a dynamic analysis tool such as @MASTG-TOOL-0038 on the app and execute its [automated jailbreak detection bypass commands](../../../Document/0x06j-Testing-Resiliency-Against-Reverse-Engineering.md#automated-jailbreak-detection-bypass)

## Observation

The output should include any instances of common jailbreak detection checks in the app binary and the results of the automated jailbreak detection bypass commands.

## Evaluation

> The test passes if jailbreak detection is implemented and the automated jailbreak detection bypass commands are successful or at least provide useful information to help you narrow down the jailbreak detection checks implemented in the app.

Reviewer 2: ```suggestion
The test passes if the automated jailbreak detection bypass command indicates that the application is testing for known jailbreak artefacts.
```

The test fails if jailbreak detection is not implemented. However, note that this test is not exhaustive and may not detect all jailbreak detection checks as it relies on predefined bypass code that may not cover all possible jailbreak detection checks or may not be up-to-date. The checks may also be more sophisticated than what the tool can detect so manual reverse engineering and deobfuscation may be required to identify them.

> Consider that there may be other resiliency mechanisms in place that prevent the application from running, such as runtime integrity checks, so the app may still not run even if the jailbreak detection checks are bypassed, or you may not be able to get to the point where the jailbreak detection checks are executed before the app crashes.

Reviewer 2: ```suggestion
```
Would delete this. The test only requires JB detection to be implemented so we should focus on that. Even if the app only does one very basic test which is easily bypassed, this test should succeed. Whether or not you can eventually run the app is irrelevant.

> Even if the automated jailbreak detection bypass commands are not successful, they may provide useful information to help you narrow down the jailbreak detection checks implemented in the app.

Reviewer 2: ```suggestion
```
If they give any useful info, the check would already pass.
